### PR TITLE
RGW/STS: adding role ownership and verify permission policy given a role action

### PR DESCRIPTION
**Pull request by BBoozmen**
Fixes https://tracker.ceph.com/issues/63751
Issue reproducer is added at https://github.com/ceph/s3-tests/pull/535 - the scenarios leading to the issue are provided there. To sum up, currently, a random s3 user can update/delete another user's role without given any permission policies which is a critical privacy/security gap.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by pritha-srivastava** [comment]
@BBoozmen : I want to understand why should a `role` be owned by a user in rgw? A role is an entity similar to a user, but is not owned by any user.

**Comment by BBoozmen** [comment]
> @BBoozmen : I want to understand why should a `role` be owned by a user in rgw? A role is an entity similar to a user, but is not owned by any user. A role in rgw is based on an IAM role in AWS.

In AWS, there's this "Account" concept but in RGW world there's no such thing but only rgw-user (tenancy). That's the reason we believe a role can be associated with the s3 user (rgw tenancy) who initiated the "create role" action.

As the added integration tests indicate, otherwise, any s3 user can do anything with any role which is a huge security/privacy gap.

**Comment by BBoozmen** [comment]
> If that is the case then the user has the following caps: 'roles=*'.

Yes, the "iam" and "alt" user have this capability set in the test.

So, what you're suggesting is -- if all existing rgw users are given `"roles=*"` capabilities -- any of them can do anything with any other user's role?

**Comment by BBoozmen** [comment]
> can you clarify whether that s3 user needs the "roles" admin capability to read/write roles? it's assumed that normal users won't have those admin permissions

Hmm, let me circle back with our Product team who are trying to integrate IAM/STS to our clients' workflow and to the best of my knowledge in the use cases they test s3 clients can create roles. That's how this issue was brought to us.

> there is work in progress for iam accounts in https://github.com/ceph/ceph/pull/54333 that will allow roles to owned by accounts, and limit permissions to the owning account. that implementation will look fairly similar to what you've done here

Ahh, interesting. Let me take a look.

**Comment by pritha-srivastava** [comment]
> > @BBoozmen : I want to understand why should a `role` be owned by a user in rgw? A role is an entity similar to a user, but is not owned by any user. A role in rgw is based on an IAM role in AWS.
> 
> In AWS, there's this "Account" concept but in RGW world there's no such thing but only rgw-user (tenancy). That's the reason we believe a role can be associated with the s3 user (rgw tenancy) who initiated the "create role" action.
> 

Yes, currently a tenant maps roughly to an account. But as Casey said work on integrating accounts is underway.
> As the added integration tests indicate, otherwise, any s3 user can do anything with any role which is a huge security/privacy gap.
Its not that any s3 user can do anything with a role, only a user that has caps for roles=*, can do anything to a role in that particular tenant.

**Comment by BBoozmen** [comment]
Hi @pritha-srivastava , @cbodley,

I wanted to have a chat with our Product team before making an update. Here's our understanding of the issue.

AFAIU, the ultimate road map for radosgw would be something like [AWS Organizations terminology and concepts](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html). However, at this point (in Reef), rgw doesn't have an "_Account_" concept (yet) but only "rgw user" (equivalent of "member account" - at least for now).

@pritha-srivastava 
>  I want to understand why should a role be owned by a user in rgw?
>
> Why would a system administrator want to give caps=roles=* to all users?

In AWS case, a client (member account) would login as AWS account administrator and it's the natural owner of the account resources including iam users, roles, etc. In our workflow, we treat each rgw-user (we also call it `tenancy`) as "member account" and thus provide both read and write rgw role capabilities (`roles=*`).  Each of our clients would be able to create roles, assume some other user's roles, if other user allows in permission policy update other user's role, etc. Otherwise, we don't have certain rgw-users acting as administrators and others with only read capability. Like an AWS member account would own its resources such as roles, currently an rgw-user would own the roles created using its credentials. In a typical scenario, any of our users would do

<pre>
>>> iam_client1 = boto3.client('iam', aws_access_key_id='<acckey>', aws_secret_access_key='<seckey>', endpoint_url=endpoint, region_name='')
>>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user1role1')
</pre>

I think it's natural the user identified with the given creds owning the role being created. If/When we have the "account" concept and this user belongs to an account, then the associated account would be the owner and the user would be acting like a true "iam user".

@cbodley 
> there is work in progress for iam accounts in https://github.com/ceph/ceph/pull/54333 that will allow roles to owned by accounts, and limit permissions to the owning account.

Even when we introduce "Account" concept and be able to create iam users (~ rgw users) within an account, we still need role ownership, right? We still need the change introduced here but instead of assigning `user_id` as the owner, we'd (if the user has been assigned/migrated to an account) use `account_id` instead? I mean we'd still need a sense of role (which is now an account resource) ownership.

More importantly, even when we have account concept, the issue being addressed in this PR will still exist. All our member accounts (clients) will still need both read/write roles capabilities (`roles=*`) to be able to create their roles, assume and (if necessary perms are given) modify some other account's roles, etc.  I mean the change/fix for https://github.com/ceph/ceph/blob/cc1951a5d6d7a11185fc095b6ccd094f3b1273eb/src/rgw/rgw_rest_role.cc#L94 will still be needed -- it shouldn't be enough just to have read/write rgw roles caps to skip permission policy checks.

@cbodley 
> do folks think it's separately useful to allow non-account users to own roles that any other user can assume?

I think this is inevitable to allow backward compatibility, right? After an upgrade to a version that supports "account", until we fully migrate our clients to accounts, there'll be a mix of account-users and non-account users, right?

>  there is work in progress for iam accounts in https://github.com/ceph/ceph/pull/54333

BTW, is there any timeline for the general availability of this feature? Will it be available in one of the upcoming Reef minor versions?

To sum up, to me, this change and https://github.com/ceph/ceph/pull/54333 are more of complementary changes

- Role ownership: while here user_id is picked as the owner of a role, once account concept is introduced, either user_id or account_id (if the user is associated with an account) would be used as the owner.
- and, regardless of ownership discussion, we still need to verify permission policies after account's or user's rgw caps are satisfied -- in the end, all accounts would have `roles=*` caps.

**Comment by pritha-srivastava** [comment]
> Hi @pritha-srivastava , @cbodley,
> 
> I wanted to have a chat with our Product team before making an update. Here's our understanding of the issue.
> 
> AFAIU, the ultimate road map for radosgw would be something like [AWS Organizations terminology and concepts](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html). However, at this point (in Reef), rgw doesn't have an "_Account_" concept (yet) but only "rgw user" (equivalent of "member account" - at least for now).
> 
@cbodley : Doesnt a tenant map roughly to an AWS account for us ( and not a rgw user)
> @pritha-srivastava
> 
> > I want to understand why should a role be owned by a user in rgw?
> > Why would a system administrator want to give caps=roles=* to all users?
> 
> In AWS case, a client (member account) would login as AWS account administrator and it's the natural owner of the account resources including iam users, roles, etc. In our workflow, we treat each rgw-user (we also call it `tenancy`) as "member account" and thus provide both read and write rgw role capabilities (`roles=*`). Each of our clients would be able to create roles, assume some other user's roles, if other user allows in permission policy update other user's role, etc. Otherwise, we don't have certain rgw-users acting as administrators and others with only read capability. Like an AWS member account would own its resources such as roles, currently an rgw-user would own the roles created using its credentials. In a typical scenario, any of our users would do
> 
> >>> iam_client1 = boto3.client('iam', aws_access_key_id='', aws_secret_access_key='', endpoint_url=endpoint, region_name='')
> >>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user1role1')
> I think it's natural the user identified with the given creds owning the role being created. If/When we have the "account" concept and this user belongs to an account, then the associated account would be the owner and the user would be acting like a true "iam user".
> 
Can you clarify this part - `we treat each rgw-user (we also call it `tenancy`) as "member account" and thus provide both read and write rgw role capabilities (`roles=*`).`, the part where you say each rgw-user is called as `tenancy`. A tenant is clearly different from a rgw user.

> @cbodley
> 
> > there is work in progress for iam accounts in #54333 that will allow roles to owned by accounts, and limit permissions to the owning account.
> 
> Even when we introduce "Account" concept and be able to create iam users (~ rgw users) within an account, we still need role ownership, right? We still need the change introduced here but instead of assigning `user_id` as the owner, we'd (if the user has been assigned/migrated to an account) use `account_id` instead? I mean we'd still need a sense of role (which is now an account resource) ownership.
> 
> More importantly, even when we have account concept, the issue being addressed in this PR will still exist. All our member accounts (clients) will still need both read/write roles capabilities (`roles=*`) to be able to create their roles, assume and (if necessary perms are given) modify some other account's roles, etc. I mean the change/fix for
> 
> https://github.com/ceph/ceph/blob/cc1951a5d6d7a11185fc095b6ccd094f3b1273eb/src/rgw/rgw_rest_role.cc#L94
> 
> will still be needed -- it shouldn't be enough just to have read/write rgw roles caps to skip permission policy checks.
Can you explain why it wouldn't be sufficient to have read/write caps for roles to skip permission policy checks. The whole idea was to assign admin capabilities to a user to be able to create/update roles.
> @cbodley
> 
> > do folks think it's separately useful to allow non-account users to own roles that any other user can assume?
> 
> I think this is inevitable to allow backward compatibility, right? After an upgrade to a version that supports "account", until we fully migrate our clients to accounts, there'll be a mix of account-users and non-account users, right?
> 
> > there is work in progress for iam accounts in #54333
> 
> BTW, is there any timeline for the general availability of this feature? Will it be available in one of the upcoming Reef minor versions?
> 
> To sum up, to me, this change and #54333 are more of complementary changes
> 
> * Role ownership: while here user_id is picked as the owner of a role, once account concept is introduced, either user_id or account_id (if the user is associated with an account) would be used as the owner.
> * and, regardless of ownership discussion, we still need to verify permission policies after account's or user's rgw caps are satisfied -- in the end, all accounts would have `roles=*` caps.

Can you give me an example of a scenario where verification of policies will be needed for a user despite it having rgw caps set to roles=*

**Comment by BBoozmen** [comment]
@pritha-srivastava 
> @cbodley : Doesnt a tenant map roughly to an AWS account for us ( and not a rgw user)
> ...
> Can you clarify this part - we treat each rgw-user (we also call it tenancy) as "member account" and thus provide both read and write rgw role capabilities (roles=*)., the part where you say each rgw-user is called as tenancy. A tenant is clearly different from a rgw user.

the statement `we also call it tenancy` is more of an internal reference that we use - otherwise, we don't use any `tenant` concept as in a `multi-tenant` system. You can simply ignore that comment. Sorry for the confusion. We basically treat each rgw-user as a member-account (to make an analogy to [AWS Organizations terminology and concepts](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html)).

@pritha-srivastava 
> Can you give me an example of a scenario where verification of policies will be needed for a user despite it having rgw caps set to roles=*

https://github.com/ceph/s3-tests/pull/535 provides 2 cases for that ... when "iam" and "alt" users are given `roles=*` caps, 2 cases fail without the changes found in this PR. Since the current implementation returns right after seeing the user has read/write rgw roles caps, 2 cases fail as follows:

The test `test_role_ownership__delete_role_without_permission` fails with
<pre>
FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__delete_role_without_permission - AssertionError: without acquiring DeleteRole permission, cannot delete other user's role
</pre>

which shows that another user can delete some other user's role.

The test `test_role_ownership__put_or_delete_policy_without_permission` fails with
<pre>
FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__put_or_delete_policy_without_permission - AssertionError: alt user put the policy although permission is not granted yet
</pre>

which shows that another user can put a policy to some other user's role without obtaining PutPolicy permission - which would be averted if permission policies were verified as well.

**Comment by pritha-srivastava** [comment]
> @pritha-srivastava
> 
> > @cbodley : Doesnt a tenant map roughly to an AWS account for us ( and not a rgw user)
> > ...
> > Can you clarify this part - we treat each rgw-user (we also call it tenancy) as "member account" and thus provide both read and write rgw role capabilities (roles=*)., the part where you say each rgw-user is called as tenancy. A tenant is clearly different from a rgw user.
> 
> the statement `we also call it tenancy` is more of an internal reference that we use - otherwise, we don't use any `tenant` concept as in a `multi-tenant` system. You can simply ignore that comment. Sorry for the confusion. We basically treat each rgw-user as a member-account (to make an analogy to [AWS Organizations terminology and concepts](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html)).
> 
> @pritha-srivastava
> 
> > Can you give me an example of a scenario where verification of policies will be needed for a user despite it having rgw caps set to roles=*
> 
> [ceph/s3-tests#535](https://github.com/ceph/s3-tests/pull/535) provides 2 cases for that ... when "iam" and "alt" users are given `roles=*` caps, 2 cases fail without the changes found in this PR. Since the current implementation returns right after seeing the user has read/write rgw roles caps, 2 cases fail as follows:
> 
> The test `test_role_ownership__delete_role_without_permission` fails with
> 
> FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__delete_role_without_permission - AssertionError: without acquiring DeleteRole permission, cannot delete other user's role
> which shows that another user can delete some other user's role.
> 
> The test `test_role_ownership__put_or_delete_policy_without_permission` fails with
> 
> FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__put_or_delete_policy_without_permission - AssertionError: alt user put the policy although permission is not granted yet
> which shows that another user can put a policy to some other user's role without obtaining PutPolicy permission - which would be averted if permission policies were verified as well.

But is that is the whole point of assigning caps to a user (according to the design) - to override the permission policy - I think that the use case that you have described here is not valid.

**Comment by pritha-srivastava** [comment]
@cbodley : While there is work underway to introduce `accounts`, is there any plan to introduce the concept of a root user in rgw?

**Comment by BBoozmen** [comment]
> But is that is the whole point of assigning caps to a user (according to the design) - to override the permission policy - I think that the use case that you have described here is not valid.

Hmm, I'm not sure whether I'm on the same page - we can perhaps discuss this in one of the upcoming rgw refactoring meetings. But, without caps, a client can't even create a role:

1. Create a user
<pre>
user create --display-name=myuser1 --uid=myuser1 --access-key=acckey --secret-key=seckey
</pre>

2. Create IAM client for this user
<pre>
iam_client1 = boto3.client('iam', aws_access_key_id='acckey', aws_secret_access_key='seckey', endpoint_url=endpoint, region_name='')
</pre>

3. Create a role fails - since no `roles=write` **rgw** caps is assigned yet to user 1
<pre>
>>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user2CanAssumeThisRole')
..
botocore.exceptions.ClientError: An error occurred (Unknown) when calling the CreateRole operation: Unknown
</pre>

rgw event showing 403 resp:
<pre>
2023-12-12T23:53:02.842-0500 7fae829a1700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser1 [12/Dec/2023:23:53:02.840 -0500] "POST / HTTP/1.1" 403 206 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.002000008s
</pre>

4. Need to give write cap to this user so that it can create a role
<pre>
caps add --uid="myuser1" --caps="roles=write"
</pre>

5. Now, user can create the role
<pre>
>>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user2CanAssumeThisRole')
>>> role_response1['ResponseMetadata']['HTTPStatusCode']
200
</pre>

rgw event:
<pre>
2023-12-12T23:56:42.856-0500 7fae7898d700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser1 [12/Dec/2023:23:56:42.843 -0500] "POST / HTTP/1.1" 200 744 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.013000054s
</pre>

6. Let's say we create a second user and give it the rgw roles:write cap as well so that it can go ahead and do the same (create its own roles)

<pre>
user create --display-name=myuser2 --uid=myuser2 --access-key=acckey2 --secret-key=seckey2
caps add --uid="myuser2" --caps="roles=write"
</pre>

7. Due to the issue I'm trying to explain here, user 2 can do whatever it wants with user 1's role without acquiring any permission (just by getting rgw roles:write cap) - e.g., delete user1's role created in Step 5.

<pre>
>>> iam_client2 = boto3.client('iam', aws_access_key_id='acckey2', aws_secret_access_key='seckey2', endpoint_url=endpoint, region_name='')

>>> iam_client2.delete_role(RoleName="user2CanAssumeThisRole")
{'ResponseMetadata': {'RequestId': 'tx0000035d4a55ae01617d1-0065793ad3-14125-default', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amz-request-id': 'tx0000035d4a55ae01617d1-0065793ad3-14125-default', 'content-type': 'application/xml', 'server': 'Ceph Object Gateway (reef)', 'content-length': '149', 'date': 'Wed, 13 Dec 2023 05:02:11 GMT', 'connection': 'Keep-Alive'}, 'RetryAttempts': 0}}
</pre>

rgw resp:
<pre>
2023-12-13T00:02:11.344-0500 7fae6c174700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser2 [13/Dec/2023:00:02:11.332 -0500] "POST / HTTP/1.1" 200 149 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.012000050s
</pre>

So, you're suggesting this is not a valid usecase?

Perhaps I'm missing something here but I don't see why we'd design it such a way that just adding rgw roles capabilities  would be used to override permission policy check.

**Comment by pritha-srivastava** [comment]
> > @pritha-srivastava
> > > @cbodley : Doesnt a tenant map roughly to an AWS account for us ( and not a rgw user)
> > > ...
> > > Can you clarify this part - we treat each rgw-user (we also call it tenancy) as "member account" and thus provide both read and write rgw role capabilities (roles=*)., the part where you say each rgw-user is called as tenancy. A tenant is clearly different from a rgw user.
> > 
> > 
> > the statement `we also call it tenancy` is more of an internal reference that we use - otherwise, we don't use any `tenant` concept as in a `multi-tenant` system. You can simply ignore that comment. Sorry for the confusion. We basically treat each rgw-user as a member-account (to make an analogy to [AWS Organizations terminology and concepts](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_getting-started_concepts.html)).
> > @pritha-srivastava
> > > Can you give me an example of a scenario where verification of policies will be needed for a user despite it having rgw caps set to roles=*
> > 
> > 
> > [ceph/s3-tests#535](https://github.com/ceph/s3-tests/pull/535) provides 2 cases for that ... when "iam" and "alt" users are given `roles=*` caps, 2 cases fail without the changes found in this PR. Since the current implementation returns right after seeing the user has read/write rgw roles caps, 2 cases fail as follows:
> > The test `test_role_ownership__delete_role_without_permission` fails with
> > FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__delete_role_without_permission - AssertionError: without acquiring DeleteRole permission, cannot delete other user's role
> > which shows that another user can delete some other user's role.
> > The test `test_role_ownership__put_or_delete_policy_without_permission` fails with
> > FAILED s3tests_boto3/functional/test_sts.py::test_role_ownership__put_or_delete_policy_without_permission - AssertionError: alt user put the policy although permission is not granted yet
> > which shows that another user can put a policy to some other user's role without obtaining PutPolicy permission - which would be averted if permission policies were verified as well.
> 
> But is that is the whole point of assigning caps to a user (according to the design) - to override the permission policy - I think that the use case that you have described here is not valid.

I also wanted to point out that the policy that allows or doesn't allow to delete roles or role policies should be attached to the identity (iam_alt_client in https://github.com/ceph/s3-tests/pull/535/files) that should be allowed or disallowed, and not to the role itself, as is done in https://github.com/ceph/s3-tests/pull/535/files. There is a REST API for doing so is here: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam/client/put_user_policy.html.

**Comment by pritha-srivastava** [comment]
> > But is that is the whole point of assigning caps to a user (according to the design) - to override the permission policy - I think that the use case that you have described here is not valid.
> 
> Hmm, I'm not sure whether I'm on the same page - we can perhaps discuss this in one of the upcoming rgw refactoring meetings. But, without caps, a client can't even create a role:
> 
> 1. Create a user
> 
> user create --display-name=myuser1 --uid=myuser1 --access-key=acckey --secret-key=seckey
> 2. Create IAM client for this user
> 
> iam_client1 = boto3.client('iam', aws_access_key_id='acckey', aws_secret_access_key='seckey', endpoint_url=endpoint, region_name='')
> 3. Create a role fails - since no `roles=write` **rgw** caps is assigned yet to user 1
> 
> >>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user2CanAssumeThisRole')
> ..
> botocore.exceptions.ClientError: An error occurred (Unknown) when calling the CreateRole operation: Unknown
> rgw event showing 403 resp:
> 
> 2023-12-12T23:53:02.842-0500 7fae829a1700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser1 [12/Dec/2023:23:53:02.840 -0500] "POST / HTTP/1.1" 403 206 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.002000008s
> 4. Need to give write cap to this user so that it can create a role
> 
> caps add --uid="myuser1" --caps="roles=write"
> 5. Now, user can create the role
> 
> >>> role_response1 = iam_client1.create_role(AssumeRolePolicyDocument=policy_doc_let_assume, Path='/', RoleName='user2CanAssumeThisRole')
> >>> role_response1['ResponseMetadata']['HTTPStatusCode']
> 200
> rgw event:
> 
> 2023-12-12T23:56:42.856-0500 7fae7898d700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser1 [12/Dec/2023:23:56:42.843 -0500] "POST / HTTP/1.1" 200 744 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.013000054s
> 6. Let's say we create a second user and give it the rgw roles:write cap as well so that it can go ahead and do the same (create its own roles)
> 
> user create --display-name=myuser2 --uid=myuser2 --access-key=acckey2 --secret-key=seckey2
> caps add --uid="myuser2" --caps="roles=write"
> 7. Due to the issue I'm trying to explain here, user 2 can do whatever it wants with user 1's role without acquiring any permission (just by getting rgw roles:write cap) - e.g., delete user1's role created in Step 5.
> 
> >>> iam_client2 = boto3.client('iam', aws_access_key_id='acckey2', aws_secret_access_key='seckey2', endpoint_url=endpoint, region_name='')
> 
> >>> iam_client2.delete_role(RoleName="user2CanAssumeThisRole")
> {'ResponseMetadata': {'RequestId': 'tx0000035d4a55ae01617d1-0065793ad3-14125-default', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amz-request-id': 'tx0000035d4a55ae01617d1-0065793ad3-14125-default', 'content-type': 'application/xml', 'server': 'Ceph Object Gateway (reef)', 'content-length': '149', 'date': 'Wed, 13 Dec 2023 05:02:11 GMT', 'connection': 'Keep-Alive'}, 'RetryAttempts': 0}}
> rgw resp:
> 
> 2023-12-13T00:02:11.344-0500 7fae6c174700  1 beast: 0x7fadb900d730: 127.0.0.1 - myuser2 [13/Dec/2023:00:02:11.332 -0500] "POST / HTTP/1.1" 200 149 - "Boto3/1.28.70 md/Botocore#1.31.70 ua/2.0 os/linux#4.18.0-477.21.1.el8_8.x86_64 md/arch#x86_64 lang/python#3.11.2 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.31.70" - latency=0.012000050s
> So, you're suggesting this is not a valid usecase?
> 
> Perhaps I'm missing something here but I don't see why we'd design it such a way that just adding rgw roles capabilities would be used to override permission policy check.

Like I explained earlier - the caps functionality was added to give root capabilities to a user.

Now if you want to allow/ disallow a user to create a role/ policy, then you can do that by attaching an IAM policy to the user for doing so. And for a user to be able to attach a policy to another user should again have caps set to user-policy=*.

**Comment by cbodley** [comment]
> @cbodley : While there is work underway to introduce `accounts`, is there any plan to introduce the concept of a root user in rgw?

@pritha-srivastava yes, the account root user is mentioned in the [account design doc](https://github.com/ceph/ceph/pull/54045/files) and in the [feature's TODO list](https://github.com/ceph/ceph/pull/54333):
> IAM APIs
>- [ ] define the permission model for the account 'root user' - 'bool root' flag in RGWUserInfo similar to 'system' and 'admin'? or a separate 'iam' cap for granular read/write permission?
>- [ ] adapt existing Role APIs to account permission model (but still need to support existing 'global' roles)

the service provider/admin would create an account and its root user for a customer, and the customer would use that account root user to manage the account's users/roles without needing the global admin caps

so the ultimate goal of the account feature is to give end-users the ability to manage their own users/roles. the admin capability model has always been intended for admins only - which is why they provide read/write access to all roles - so i don't want to end up with a model that expects end-users to have those caps

**Comment by cbodley** [comment]
> I think this is inevitable to allow backward compatibility, right? After an upgrade to a version that supports "account", until we fully migrate our clients to accounts, there'll be a mix of account-users and non-account users, right?

@BBoozmen my plan was to leave existing roles as global and only accessible by an admin persona. roles created by an account (or existing roles migrated into a new account by an admin) would be accessible to that account's root user

the main part of role ownership is isolating the owner's roles from other owners. i implemented the per-owner tracking of such resources in:

> - [ ] extend cls_user to track {account}.users, {account}.roles, {account}.groups (started in https://github.com/ceph/ceph/pull/54563)

we could potentially use that to implement user ownership via a {userid}.roles object, but i'm not sure that extra work is worth the effort...

> BTW, is there any timeline for the general availability of this feature? Will it be available in one of the upcoming Reef minor versions?

i've been hoping it could make the squid release, but there's a lot to do and i'd need help to get everything working by then. whenever that lands, we wouldn't plan to do backports. similarly, i don't think we'd consider backporting intrusive changes to user-owned roles either
